purple-pill/js/controllers/fontexportController.js

```diff
@@ -20,152 +92,136 @@ function($scope, $http, sharedScope, $timeout) {
+        hours = zeroPadding(date.getHours());
+        minutes = zeroPadding(date.getMinutes());
+        seconds = zeroPadding(date.getSeconds());
+
+        return [year, month, day].join("") + "-" + [hours, minutes, seconds].join("");
+    }
+
+    const UI_UPDATE_TIMESLICE = 50; // msecs
+    var exportIsRunning = false;
+
+    function exportingGlyphMessage (it, instanceIndex, totalInstances){
+        var msg
+          , currentGlyph = it.value['current_glyph'] + 1 //humans start counting from 1.
+          , totalGlyphs = it.value['total_glyphs']
+          , glyphId = it.value['glyph_id']
+          ;
+        msg = totalInstances + " instances to export."
+        msg += "Calculating glyph '" + glyphId + "' (" + currentGlyph + " of " + totalGlyphs + ")"
+        msg += " of instance #" + (instanceIndex+1);
+        return msg;
+    }
+
+    function calculateGlyphsProgress(it, instanceIndex, totalInstances){
+        var percentage
+          , currentGlyph = it.value['current_glyph']
+          , totalGlyphs = it.value['total_glyphs']
+          ;
+        percentage = 100.0 * (instanceIndex + (currentGlyph/totalGlyphs)) / totalInstances;
+        return percentage;
+    }
 
-        function get_timestamp(){
-            var year, month, day, hours, minutes, seconds
-              , date = new Date()
-              ;
-            year = zero_padding(date.getFullYear());
-            month = zero_padding(date.getMonth() + 1); //getMonth() returns a zero-based value: 0=January, 11=December
-            day = zero_padding(date.getDate());
-            hours = zero_padding(date.getHours());
-            minutes = zero_padding(date.getMinutes());
-            seconds = zero_padding(date.getSeconds());
-
-            return [year, month, day].join("") + "-" + [hours, minutes, seconds].join("");
-        }
-
+    function exportFont_compute_glyphs(exportObjects) {
         var bundle = new $scope.data.stateless.JSZip()
-          , bundleFolderName = "metapolator-export-" + get_timestamp()
-          , bundle_filename = bundleFolderName + ".zip"
+          , bundleFolderName = "metapolator-export-" + getTimestamp()
+          , bundleFileName = bundleFolderName + ".zip"
           , bundleFolder = bundle.folder(bundleFolderName)
-          , bundle_data
-          , instances_for_export = Array()
+          , totalInstances = exportObjects.length
+          , progress = new ProgressBar( $("#progressbar")
+                                      , $("#progresslabel")
+                                      , UI_UPDATE_TIMESLICE )
           ;
-        angular.forEach($scope.data.families, function(family) {
-            angular.forEach(family.instances, function(instance) {
-                if (instance.exportFont){
-                    instances_for_export.push(instance);
-                }
-            });
-        });
+        function _exportFontComputeGlyphs(exportObjects, totalInstances, bundleFolder, resolve, reject){
```

Comment:
Since this function is exclusively in the scope of `exportFont_compute_glyphs` and since this is getting much clearer to overview, I think we could remove the three first arguments here and indeed use the names from the closure.
Alternatively, we must add one more argument `progress` which we forgot to add before. The latter is better portable, of course. (meaning that this function could be moved easily out of it's closure without having to rewrite it).